src/x/checked/finalizeable_once.go

```diff
@@ -0,0 +1,18 @@
+package checked
+
+import "go.uber.org/atomic"
+
+// FinalizeableOnce implements the logics needed to enforce single Finalize on pool.ObjectPool.
+type FinalizeableOnce struct {
+	finalized     atomic.Bool
+}
+
+// Finalized returns true iff the object has been finalized.
+func (c *FinalizeableOnce) Finalized() bool {
+	return c.finalized.Load()
+}
+
+// SetFinalized sets the flag of object finalize.
+func (c *FinalizeableOnce) SetFinalized(f bool) {
+	c.finalized.Store(f)
```

Comment:
I would use the underlying "Swap(value)" here and return what the old value was.

That way we can atomically see if we transitioned from false to true.

i.e. 
```go
func (c *FinalizeableOnce) SetFinalized() error {
  if oldValue := c.finalized.Swap(true); oldValue == true {
    return errAlreadyFinalized
  }
  return nil
}

func (c *FinalizeableOnce) SetUnfinalized() error {
  if oldValue := c.finalized.Swap(false); oldValue == false {
    return errNotFinalized
  }
  return nil
}
```

Reply:
Nice idea, changing.